src/Agent.Plugins/PipelineCache/PipelineCacheServer.cs

```diff
@@ -153,26 +155,106 @@ public class PipelineCacheServer
+                process.StartInfo.FileName = @"C:\Program Files\7-Zip\7z.exe"; // tar // @"C:\Program Files\7-Zip\7z.exe"
+                process.StartInfo.Arguments = $"x -si -aoa -o{targetDirectory} -ttar"; // -xf - -C {targetDirectory} // tarring.StartInfo.Arguments = $"x -si -aoa -o{targetDirectory} -ttar";
+                process.StartInfo.UseShellExecute = false;
+                process.StartInfo.RedirectStandardInput = true;
+                process.StartInfo.RedirectStandardOutput = true;
+                process.StartInfo.RedirectStandardError = true;
+                process.EnableRaisingEvents = true;
+                process.Exited += (sender, args) =>
+                {
+                    cancelSource.Cancel();
+                    processTcs.SetResult(process.ExitCode);
+                };
+
+                context.Info($"Starting '{process.StartInfo.FileName}' with arguments '{process.StartInfo.Arguments}'...");
+                process.Start();
+
+                var output = new List<string>();
+
+                Func<string, StreamReader, Task> readLines = (prefix, reader) => Task.Run(async() => {
+                    string line;
+                    while(null != (line = await reader.ReadLineAsync()))
+                    {
+                        lock(output)
+                        {
+                            output.Add($"{prefix}{line}");
+                        }
+                    }
+                });
+
+                Task readStdOut = readLines("stdout: ", process.StandardOutput);
+                Task readStdError = readLines("stderr: ", process.StandardError);
+                Task downloadTask = Task.Run(async () => {
+                    try
+                    {
+                        await dedupManifestClient.DownloadPipelineCacheAsync(options, linkedSource.Token, process.StandardInput.BaseStream);
+                        process.StandardInput.BaseStream.Close();
+                    }
+                    catch (Exception e)
+                    {
+                        process.Kill();
+                        ExceptionDispatchInfo.Capture(e).Throw();
+                    }
+                });
+
+                // Our goal is to always have the process ended or killed by the time we exit the function.
+
+                try
+                {
+                    using (cancellationToken.Register(() => process.Kill()))
+                    {
+                        // readStdOut and readStdError should only fail if the process dies
+                        // processTcs.Task cannot fail as we only call SetResult on processTcs
+                        // downloadTask *can* fail, but when it does, it will also kill the process
+                        await Task.WhenAll(readStdOut, readStdError, processTcs.Task, downloadTask);
+                    }
+
+                    int exitCode = await processTcs.Task;
+
+                    if (exitCode == 0)
```

Comment:
Any reason we're not always showing stdout/stderr from the process? (i.e. leave `RedirectStandardOutput` set to `false`)?

Alternatively, I would expect to show stdout/stderr when `exitCode !=` since the user (or us) will want to understand why the process failed.